# ECO462: Homework 4

In [77]:
# Import basic Python libraries
import pandas as pd 
import numpy as np 
from scipy import stats
import statsmodels.api as sm

## Question 1: Properties of Fama-French’s Five Factors

In [78]:
# Question 1a Coding Portion

df_factors = pd.read_excel("../data/HW4 Data.xlsx", "Factors")
df_factors["Date"] = pd.to_datetime(df_factors["Date"])
df_factors = df_factors.set_index("Date")

df_factors = df_factors / 100

print("\n================== Summary Statistics ==================\n")
print(df_factors.describe())
print("\n================== Correlation Matrix ==================\n")
print(df_factors.corr())


================== Summary Statistics ==================

           Mkt-RF         SMB         HML         RMW         CMA          RF
count  733.000000  733.000000  733.000000  733.000000  733.000000  733.000000
mean     0.005795    0.002048    0.002873    0.002886    0.002606    0.000036
std      0.044868    0.030482    0.030001    0.022196    0.020760    0.000026
min     -0.232400   -0.153200   -0.138800   -0.186500   -0.072000    0.000000
25%     -0.019700   -0.015600   -0.014000   -0.007800   -0.010200    0.000014
50%      0.009300    0.000700    0.002200    0.002600    0.000900    0.000038
75%      0.034400    0.020100    0.017500    0.013200    0.014700    0.000050
max      0.161000    0.182800    0.128000    0.130700    0.090700    0.000135

================== Correlation Matrix ==================

          Mkt-RF       SMB       HML       RMW       CMA        RF
Mkt-RF  1.000000  0.279300 -0.204133 -0.186581 -0.361344 -0.083120
SMB     0.279300  1.000000  0.001820 -0.352176

### Question 1b Written Portion

**Question:** *Let’s focus on the High-Minus-Low (HML) factor for a moment. How would you
construct a mimicking portfolio that replicates it? Why would you expect this
portfolio to move one-to-one with HML and not with the other factors?*

**Answer**: A mimicking portfolio for the High-Minus-Low (HML) factor, as defined in the Fama–French factor framework, is constructed by taking long and short positions in portfolios sorted by book-to-market ratios. Specifically, assets are typically grouped into three quantiles (high, medium, and low) based on their book-to-market values. The HML factor is then formed by going long high book-to-market (“value”) stocks and short low book-to-market (“growth”) stocks, generally with equal weights across each leg.

In the broader factor-modeling framework, mimicking portfolios are designed to isolate exposure to a single systematic factor. In the first-pass regression, each asset’s factor loadings (betas) are estimated by regressing excess returns on the factor returns. Suppose there are $n$ assets and $k$ factors, yielding a beta matrix $\beta \in \mathbb{R}^{n\times k}$. The goal is to find portfolio weights $w \in \mathbb{R}^{n}$ such that:
$$B^{\intercal}w = e_j$$
where $e_j$ is the unit vector corresponding to the HML factor (or any arbitrary target factor at index $j$), implying the portfolio has beta 1 with respect to HML and 0 with respect to all other factors.

Given that $n>k$ and the column of $B$ are linearly independent, there exists at least one $w$ that satisfies the system, ensuring that the mimicking portfolio's returns move 1-for-1 with HML and are orthogonal to the other factors.The mimicking portfolio is explicitly constructed to neutralize any exposure to all non-HML factors, thus any variation can be solely attributed to the HML factor and therefore does not covary with any other factor.

In [79]:
## Question 1c Coding Portion

t_statistic, p_value = stats.ttest_1samp(df_factors["HML"], 0)
print("\n================== T-Test Results ==================")
print(f"t-statistic: {t_statistic:.2f}\np-value: {p_value:.2f}")


================== T-Test Results ==================
t-statistic: 2.59
p-value: 0.01


### Question 1c Written Portion

**Question**: Based on the moments of the HML factor, run a simple test to determine whether
high book-to-market firms have, on average, higher returns than low book-to-
market firms during the sample period. Hint: How can you test whether the
HML factor’s mean is statistically different from zero?

To test whether high book-to-market firms earn higher returns on average compared to low book-to-market firms, we perform a one-sample t-test on the HML factor's mean. The null hypothesis is that the HML factor's mean is equal to 0 ($H_0: \mu _{HML} = 0$), implying no difference in return on average between the high and low book-to-market firms. The alternative hypothesis is that high book-to-market firms earn higher average returns ($H_A:\mu _{HML} > 0$).

Using a significance level of 5% ($\alpha = 0.05$), the t-test compares the sample mean of the HML factor to zero. The resulting p-value of **0.01** is below the significance threshold, providing strong evidence to reject the null hypothesis and accept the alternative hypothesis. The positive t-statistic further indicates the mean of the HML factor is greater than 0.

Therefore, we can conclude that during the sample period, high book-to-market firms earned significantly higher average returns than low book-to-market firms.

## Question 2: Regression Using Fama-French's Three Factors

In [80]:
# Question 2a Coding Portion

factors = df_factors.columns.tolist()

df_stock_returns = pd.read_excel("../data/HW4 Data.xlsx", "Stock Returns")
df_stock_returns["DATE"] = pd.to_datetime(df_stock_returns["DATE"])
df_stock_returns = df_stock_returns.set_index("DATE")

tickers = df_stock_returns.columns.tolist()
df_results = pd.DataFrame()

def factor_model(t, factors):
  df_asset = pd.DataFrame(index=df_stock_returns.index)
  df_asset[t+"_Excess_Returns"] = df_stock_returns[t] - df_factors["RF"]
  for factor in factors[:-3]:
    df_asset[t+"_"+factor] = df_factors[factor]
  return df_asset

for t in tickers:
  df_stock = factor_model(t, factors)

  X = df_stock.iloc[:, 1:]
  X = sm.add_constant(X)
  y = df_stock[t+"_Excess_Returns"]

  model = sm.OLS(y, X).fit()

  t_beta_MKT = model.t_test(t+"_"+"Mkt-RF")
  t_beta_SMB = model.t_test(t+"_"+"SMB")
  t_beta_HML = model.t_test(t+"_"+"HML")

  row = pd.DataFrame({
        "Stock": [t],
        "Alpha": [model.params["const"]],
        "p-value_Alpha": [round(float(model.pvalues["const"]),5)],
        "Beta_MKT": [model.params[t+"_"+"Mkt-RF"]],
        "t-Beta_MKT": [round(float(t_beta_MKT.tvalue.item()),5)],
        "p_Beta_MKT": [round(float(t_beta_MKT.pvalue.item()),5)],
        "Beta_SMB": [model.params[t+"_"+"SMB"]],
        "t-Beta_SMB": [round(float(t_beta_SMB.tvalue.item()),5)],
        "p_Beta_SMB": [round(float(t_beta_SMB.pvalue.item()),5)],
        "Beta_HML": [model.params[t+"_"+"HML"]],
        "t-Beta_HML": [round(float(t_beta_HML.tvalue.item()),5)],
        "p_Beta_HML": [round(float(t_beta_HML.pvalue.item()),5)],
        "R²": model.rsquared
    })
  
  df_results = pd.concat([df_results,row])

df_results = df_results.set_index("Stock")
print(df_results)

          Alpha  p-value_Alpha  Beta_MKT  t-Beta_MKT  p_Beta_MKT  Beta_SMB  \
Stock                                                                        
AEP    0.004122        0.02498  0.580018    13.39122         0.0 -0.321953   
KMB    0.004075        0.04550  0.720385    15.00613         0.0 -0.380247   
BMY    0.007462        0.00080  0.830190    15.87444         0.0 -0.362140   
XOM    0.005062        0.00220  0.761535    19.58344         0.0 -0.266676   
IBM    0.004646        0.03083  0.892604    17.61031         0.0 -0.179953   

       t-Beta_SMB  p_Beta_SMB  Beta_HML  t-Beta_HML  p_Beta_HML        R²  
Stock                                                                      
AEP      -5.18203     0.00000  0.542401     8.51194     0.00000  0.235546  
KMB      -5.52202     0.00000  0.241420     3.41828     0.00067  0.245357  
BMY      -4.82754     0.00000 -0.092469    -1.20184     0.22983  0.282065  
XOM      -4.78092     0.00000  0.380721     6.65480     0.00000  0.358738

### Question 2a Written Portion

**Question:** Report your estimates for all factor loadings from the first-pass regression.
Before running the regressions, you need to construct excess returns on each of
the five stocks. Interpret the results. Are the factor loadings estimated precisely?

To assess the precision of the estimated factor loadings from the first-pass Fama-French 3-factor regressions, we perform a one-sample t-test on each coefficient for each stock ($b_i, s_i, h_i$). Formally, the null hypothesis is for each factor loading is:
$$H_0:\beta _{factor}=0$$
and the alternative hypothesis is:
$$H_A:\beta _{factor} \neq 0$$
This test evaluates whether the estimated factor loading is statistically distinguishable from zero at a conventional significance level. In the context of factor models, a rejection of the null indicates that the factor contributes meaningfully to explaining the excess return of the stock.

Using the same significance threshold of $\alpha = 0.05$, the estimated factor loadings for the market (*Beta_MKT*) and size (*Beta_SML*) are statistically significant for all give stocks, indicated by a p-values below 0.05, confirming that these factor exposures are estimated with high precision. For the value factor (*Beta_HML*), however, the factor loading was not statistically significant for BMY ($p=0.22983$), suggesting that its exposure to HML is less precisely estimated. For the other stocks, the HML factor loadings are statistically significant, supporting reasonable overall precision for this factor in the sample.

Additionally, it is important to consider the direction and magnitude of the factor exposures. Negative t-statistics for Beta_SMB indicate that for all the stocks in the provided sample are effectively "short" the size factor, meaning their returns are inversely related to the small-minus-big market-cap portfolios. Similarly, negative t-statistics for Beta_SML for some stocks suggest short exposure to the value factor, whle positive and significant t-statistics for Beta_MKT confirm that all stocks maintain a long exposure to the market factor.

Overall, the first-pass regression indicate that most factor loadings are estimated precisely, particulaly for the the market and size factors. The HML factor shows mixed precision across stocks, with one non-significant coefficient, but the pattern of significance across the remaining stocks provides sufficient evidence that the factor model captures systematic variation in excess returns effectively.

### Question 2b Written Portion

**Question**: Which stock is most exposed to each factor? If you believe that the HML factor
will outperform other factors in the near future, which of the five stocks would
you buy? Suppose that HML outperformance is accompanied by a better-than-
expected market return and a worse-than-expected SMB return. Would your
stock pick change?

Based on the first pass Fama-French 3-factor regression estimates:

*Beta_MKT*: IBM	(Factor Loading: 0.892604)

*Beta_SMB*: KMB	(Factor Loading: -0.380247)

*Beta_HML*: AEP	(Factor Loading: 0.542401)

Assume exposure refers to the magnitiude of the coefficient. IBM has the largest exposure to the market factor, indicating the most sensitivity to overall market returns. KMB has the largest magnitude of the coefficient and is therefore most exposes to SMB movements. AEP has the largest exposure to the value factor, indicating the most sensitivity to HML movements.

If HML is expected to outperform other factors, AEP is the preferred stock, since it has the highest positive Beta_HML (0.542401). 

If HML outperformance occurs alongside better-than-expected market returns and worse-than-expected SMB returns, the combined expected excess return for each stock would be approximately:

$$\mathbb{E}[R_{it}-R_{Ft}]=a_i+b_i\Delta MKT+s_i\Delta SMB+h_i\Delta HML$$

All five stocks are long the market (positive Beta_MKT) and short SMB (negative Beta_SMB). Stocks with larger exposures to these factors could generate higher total excess returns than AEP under certain scenarios. For instance, KMB has a larger market exposure than AEP (0.720385) and the largest short SMB exposure (-0.380247). If the magnitude of unexpected market and SMB movements exceeds the anticipated gain from HML, KMB could become more attractive than AEP. However, if HML is expected to strongly outperform the other factors, AEP remains the preferred choice due to its highest Beta_HML.


In [81]:
# Question 2c Coding Portion

print(f"Average R²: {df_results["R²"].mean():.4f}")

Average R²: 0.2920


### Question 2c Written Portion

**Question:** How much of the variance of the five stocks can be explained by the three factors?
What are the implications for the proportion of systematic and idiosyncratic risk
in each of the stocks? Hint: R2 measures the proportion of the variance in the
dependent variable that is explained by the independent variables in a regression.

On average, 29.2% of the total variance in stock returns is explained by the three factors, as indicated by the average R² value across the five regressions. The R² values in the results table show the proportion of each stock’s return variation that can be attributed to systematic risk, i.e. the movements driven by the common risk factors.

Since the average R² is relatively low, this implies that a large portion of each stock’s variance is idiosyncratic, or firm-specific, rather than explained by broad market, size, or value factors. In other words, only ~30% of return variability is systematic, while roughly 70% reflects unique risks specific to individual firms. This suggests that diversification across these stocks could meaningfully reduce portfolio risk, as much of their variation is not driven by common factors.


In [82]:
# Question 2d Coding Portion

df_residuals = pd.DataFrame(index=df_stock_returns.index)

for t in tickers:
    df_stock = factor_model(t, factors)
    df_residuals[t] = df_stock[t+"_"+"Excess_Returns"] - df_results["Alpha"][t] - df_results["Beta_MKT"][t] * df_factors["Mkt-RF"] - df_results["Beta_SMB"][t] * df_factors["SMB"] - df_results["Beta_HML"][t] * df_factors["HML"]

residual_corr = df_residuals.corr()
print(residual_corr)

          AEP       KMB       BMY       XOM       IBM
AEP  1.000000  0.156939  0.081206  0.032527 -0.123199
KMB  0.156939  1.000000  0.168628 -0.030139 -0.041434
BMY  0.081206  0.168628  1.000000 -0.006044 -0.004877
XOM  0.032527 -0.030139 -0.006044  1.000000  0.031987
IBM -0.123199 -0.041434 -0.004877  0.031987  1.000000


### Question 2d Written Portion

**Question:** One can calculate the correlation matrix of the regression residuals for all five
stocks. If the off-diagonal terms are far from zero, is it safe to say that the stock
that has the highest R² in the three-factor model has the highest (in proportion
to its total risk) loading on systematic risks? What argument can you use to
motivate the 5-factor model?

The off-diagonal correlations in the residual matrix are small, suggesting that the three-factor model captures most shared variation in these stocks. However, a small residual correlation does not imply that the stock with the highest R² necessarily has the greatest exposure to systematic risk. R² measures the fraction of total variance explained by included factors only, thus omitted common factors or nonlinear exposures can distort this comparison. Large residual correlations would indicate missing shared components, which is an argument for extending the model to include other factors that capture the shared variation.

In [83]:
# Question 2e Coding portion

df_historical_averages = pd.DataFrame(index=df_results.index)
for t in tickers:
  df_stock = factor_model(t, factors)
  df_historical_averages.loc[t, "Average Excess Return"] = df_stock[t+"_Excess_Returns"].mean()

factor_betas = df_results[["Beta_MKT","Beta_SMB", "Beta_HML"]]

X = sm.add_constant(factor_betas)
y = df_historical_averages["Average Excess Return"]
second_pass = sm.OLS(y, X).fit()
print(second_pass.summary())

mean_MKT = df_factors["Mkt-RF"].loc[:"2021-12-01"].mean()
mean_SMB = df_factors["SMB"].loc[:"2021-12-01"].mean()
mean_HML = df_factors["HML"].loc[:"2021-12-01"].mean()
print(second_pass.t_test(f"Beta_MKT = {mean_MKT}"))
print(second_pass.t_test(f"Beta_SMB = {mean_SMB}"))
print(second_pass.t_test(f"Beta_HML = {mean_HML}"))

                              OLS Regression Results                             
Dep. Variable:     Average Excess Return   R-squared:                       0.440
Model:                               OLS   Adj. R-squared:                 -1.240
Method:                    Least Squares   F-statistic:                    0.2619
Date:                   Fri, 17 Oct 2025   Prob (F-statistic):              0.854
Time:                           23:27:47   Log-Likelihood:                 28.103
No. Observations:                      5   AIC:                            -48.21
Df Residuals:                          1   BIC:                            -49.77
Df Model:                              3                                         
Covariance Type:               nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0041 

/opt/anaconda3/lib/python3.13/site-packages/statsmodels/stats/stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


### Question 2e Written Portion

**Question:** Estimate the factor risk premia using the second-pass regression represented
by equation (2). Expected returns, inherently forward-looking and unobservable
variables, are not directly available. How can you deal with that? What assumptions are you making? You might find that all the factor risk premia are
insignificant due to the short sample. Consider each of the three null hypotheses: Hi
0 : λi “  ̄Fi where  ̄Fi are the historical average of each factor that you
have calculated in part 1, can you reject them at the 95% confidence level? Test
separately.

In the second-pass cross-sectional regression, the historical average excess return for each stock is regressed on its estimated factor loadings from the first-pass model to obtain the factor risk premia ($\lambda _m,\lambda _s,\lambda _h$). Because expected returns are unobservable, the sample mean excess returns are used as proxies, assuming that past average performance reasonably approximates investors’ expectations—an admittedly strong stationarity assumption. The resulting coefficients are statistically insignificant (all p-values > 0.05), indicating that none of the estimated factor premia differ significantly from zero. Separate t-tests comparing each estimated λ to its corresponding historical mean factor return ($H_{0}^i:\lambda _i = \bar F_i$) also yield high p-values, so the nulls cannot be rejected at the $\alpha = 0.05$ confidence level. Given the very small cross-section of only five assets and large standard errors, these results likely reflect limited statistical power rather than evidence that the factor risk premia truly equal their historical means.

## Question 3: Regression Using Fama-French's Five Factors

In [84]:
# Question 3a Coding Portion

df_five_results = pd.DataFrame()

def five_factor_model(t, factors):
  df_asset = pd.DataFrame(index=df_stock_returns.index)
  df_asset[t+"_Excess_Returns"] = df_stock_returns[t] - df_factors["RF"]
  for factor in factors[:-1]:
    df_asset[t+"_"+factor] = df_factors[factor]
  return df_asset

for t in tickers:
  df_stock = five_factor_model(t, factors)

  X = df_stock.iloc[:, 1:]
  X = sm.add_constant(X)
  y = df_stock[t+"_Excess_Returns"]

  model = sm.OLS(y, X).fit()

  t_beta_MKT = model.t_test(t+"_Mkt-RF")
  t_beta_SMB = model.t_test(t+"_SMB")
  t_beta_HML = model.t_test(t+"_HML")
  t_beta_RMW = model.t_test(t+"_RMW")
  t_beta_CMA = model.t_test(t+"_CMA")

  row = pd.DataFrame({
        "Stock": [t],
        "Alpha": [model.params["const"]],
        "p-value_Alpha": [round(float(model.pvalues["const"]),5)],
        "Beta_MKT": [model.params[t+"_"+"Mkt-RF"]],
        "t-Beta_MKT": [round(float(t_beta_MKT.tvalue.item()),5)],
        "p_Beta_MKT": [round(float(t_beta_MKT.pvalue.item()),5)],
        "Beta_SMB": [model.params[t+"_"+"SMB"]],
        "t-Beta_SMB": [round(float(t_beta_SMB.tvalue.item()),5)],
        "p_Beta_SMB": [round(float(t_beta_SMB.pvalue.item()),5)],
        "Beta_HML": [model.params[t+"_"+"HML"]],
        "t-Beta_HML": [round(float(t_beta_HML.tvalue.item()),5)],
        "p_Beta_HML": [round(float(t_beta_HML.pvalue.item()),5)],
        "Beta_RMW": [model.params[t+"_"+"RMW"]],
        "t-Beta_RMW": [round(float(t_beta_RMW.tvalue.item()),5)],
        "p_Beta_RMW": [round(float(t_beta_RMW.pvalue.item()),5)],
        "Beta_CMA": [model.params[t+"_"+"CMA"]],
        "t-Beta_CMA": [round(float(t_beta_CMA.tvalue.item()),5)],
        "p_Beta_CMA": [round(float(t_beta_CMA.pvalue.item()),5)],
        "R²": model.rsquared
    })
  
  df_five_results = pd.concat([df_five_results,row])

df_five_results = df_five_results.set_index("Stock")
print(df_five_results)

          Alpha  p-value_Alpha  Beta_MKT  t-Beta_MKT  p_Beta_MKT  Beta_SMB  \
Stock                                                                        
AEP    0.002721        0.14829  0.630282    13.78294         0.0 -0.282744   
KMB    0.000721        0.72278  0.815269    16.48964         0.0 -0.241613   
BMY    0.004454        0.04668  0.907836    16.70137         0.0 -0.224600   
XOM    0.003716        0.02778  0.811674    19.80406         0.0 -0.232260   
IBM    0.005388        0.01504  0.857742    15.95535         0.0 -0.186115   

       t-Beta_SMB  p_Beta_SMB  Beta_HML  t-Beta_HML  p_Beta_HML  Beta_RMW  \
Stock                                                                       
AEP      -4.33093     0.00002  0.365203     4.30642     0.00002  0.153691   
KMB      -3.42302     0.00066 -0.057430    -0.62636     0.53128  0.568097   
BMY      -2.89425     0.00392 -0.323731    -3.21147     0.00138  0.568526   
XOM      -3.96942     0.00008  0.201372     2.64939     0.00825  0.1

### Question 3a Written Portion

**Question**: Estimate the Fama-French 5-factor model on five stocks. Report and interpret
the results. Compare the point estimates of the intercept from the 3 and 5-factor
models for the stock. Which stock has shown the greatest improvement in the fit
of regression (i.e., the largest increase in R2)?

Across the five stocks, the Fama-French 5-factor coefficients exhibit the expected economic patterns. The market factor (MKT) remains positive and highly significant for all firms, indicating that excess returns move closely with overall market fluctuations. The SMB loadings are consistently negative, confirming that these firms behave like large-cap stocks. HML coefficients differ in sign across firms, showing that some are more growth-oriented (negative HML) while others lean toward value (positive HML). The newly added profitability (RMW) and investment (CMA) factors are generally positive and significant, suggesting that firm-level characteristics related to profitability and investment help explain additional return variation.

Comparing model fit, the 5-factor model improves R² for every stock relative to the 3-factor specification:

AEP: 0.236 → 0.247

KMB: 0.245 → 0.293

BMY: 0.282 → 0.315

XOM: 0.359 → 0.371

IBM: 0.338 → 0.342

The largest improvement in explanatory power occurs for **KMB**, with R² increasing by approximately 0.048. This suggests that the inclusion of profitability and investment factors captures meaningful additional variation in KMB’s returns, likely reflecting the firm’s sensitivity to quality and capital allocation effects.

Regarding intercepts (alphas), their magnitudes generally decline and become less statistically significant under the 5-factor model. For example, BMY’s alpha decreases from 0.00746 (p = 0.0008) in the 3-factor model to 0.00445 (p = 0.0467) in the 5-factor model, while AEP’s alpha declines from 0.00412 (p = 0.0250) to 0.00272 (p = 0.1483). Similar reductions occur for KMB and XOM, both losing statistical significance. This pattern indicates that what appeared to be “abnormal” returns under the 3-factor framework can largely be attributed to systematic exposures captured by the additional profitability and investment factors.

### Question 3b Written Portion

**Question:** How do the a’s of each stock change from the FF 3-factor model to the five-
factor model? How can your results speak to the statement ”alpha is sometimes
undiscovered beta?”

Across all five stocks, the estimated alphas generally decline once the profitability (RMW) and investment (CMA) factors are added, and most lose statistical significance. For example, AEP’s alpha decreases from 0.0041 (p = 0.025) to 0.0027 (p = 0.148), and KMB’s drops sharply from 0.0041 (p = 0.046) to 0.0007 (p = 0.723). BMY’s alpha also falls from 0.0075 to 0.0045, becoming only marginally significant, while XOM’s declines from 0.0051 to 0.0037 with reduced significance. IBM’s alpha remains roughly the same, around 0.0054, and retains modest significance.

This overall pattern strongly supports the statement that “alpha is sometimes undiscovered beta.” Under the 3-factor model, part of what appeared to be abnormal performance likely reflected exposure to omitted sources of systematic risk. Once profitability and investment factors are included, much of the previously unexplained return is absorbed by these new betas. The disappearance or reduction of alpha therefore suggests that the so-called “abnormal” returns were not true excess performance, but rather compensation for additional risk exposures that the simpler model failed to capture.

***Statement of Collaboration (including ChatGPT)***: I collaborated with **Rosalia Mwidege** and **Theodore Ouyang**. Additionally, **ChatGPT** was used to debug any error-prone code and find the proper Excel-equivalent Python functions and libraries to properly execute the solutions to the problems, in conjunction with the hints listed on the problem set.

***Honor Code***: This assignment represents my own work in accordance with University regulations and class policy.